From dc8083b8277c5ed9e03d3a8b90ac69b06ff1ebe0 Mon Sep 17 00:00:00 2001
From: Yoni Farhadian <yonif@marvell.com>
Date: Thu, 18 Jul 2013 13:24:33 +0300
Subject: [PATCH 0847/1825] alp: PPv2: support change MTU in switch

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 91e7d49ceb7f2f80d19822b073000767f72c351b

	Change MTU to master interface (ethX that is connected to switch) will cause change MTU to
	all MUX interfaces and will change MTU in switch's ports as well

Change-Id: If0b86e17deea8fe1e420abaeacdf46b04a92c4e0
Signed-off-by: Yoni Farhadian <yonif@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2723
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Igor Patrik <igorp@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../mv_drivers_lsp/mv_mux/mv_mux_netdev.c          |   41 ++++++++++++++++----
 1 files changed, 33 insertions(+), 8 deletions(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
index 414450b..8571a9a 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
@@ -377,6 +377,22 @@ static int mv_mux_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 }
 
 /*-----------------------------------------------------------------------------------------*/
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+static void mv_mux_switch_mtu_update(int mtu)
+{
+	int pkt_size, tag_size = 0;
+	MV_TAG_TYPE tag_type =  mux_switch_shadow.tag_type;
+
+	if (tag_type == MV_TAG_TYPE_MH)
+		tag_size = MV_ETH_MH_SIZE;
+	else if (tag_type == MV_TAG_TYPE_DSA)
+		tag_size = MV_ETH_DSA_SIZE;
+
+	pkt_size = mtu + tag_size + MV_ETH_ALEN + MV_ETH_VLAN_SIZE + MV_ETH_CRC_SIZE;
+	mv_switch_jumbo_mode_set(pkt_size);
+}
+#endif
+/*-----------------------------------------------------------------------------------------*/
 
 int mv_mux_close(struct net_device *dev)
 {
@@ -414,8 +430,13 @@ int mv_mux_open(struct net_device *dev)
 	netif_stacked_transfer_operstate(root, dev);
 	netif_tx_wake_all_queues(dev);
 #ifdef CONFIG_MV_INCLUDE_SWITCH
-	if (mux_switch_shadow.attach)
+	if (mux_switch_shadow.attach) {
+		if (dev->mtu != root->mtu) {
+			dev->mtu = root->mtu;
+			mv_mux_switch_mtu_update(dev->mtu);
+		}
 		mv_switch_group_enable(pmux_priv->idx);
+	}
 #endif
 
 	printk(KERN_NOTICE "%s: started\n", dev->name);
@@ -588,6 +609,11 @@ static struct net_device *mv_mux_netdev_init(int port, struct net_device *mux_de
 
 	/*SET_ETHTOOL_OPS(mux_dev, &mv_mux_tool_ops);*/
 
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+	if (mux_switch_shadow.attach)
+		mv_mux_switch_mtu_update(mux_dev->mtu);
+#endif
+
 	return mux_dev;
 }
 /*-----------------------------------------------------------------------------------------*/
@@ -766,12 +792,10 @@ static int mux_device_event(struct notifier_block *unused, unsigned long event,
 				mv_mux_update_link(mux_dev, link_up);
 			}
 #else
-
 			/* In case of external switch, propagate real device link state to mux devices */
 			/* change state*/
 			netif_stacked_transfer_operstate(dev, mux_dev);
-
-#endif /*CONFIG_MV_INCLUDE_SWITCH*/
+#endif
 			mux_dev = pdev_priv->next;
 		}
 		break;
@@ -797,13 +821,14 @@ static int mux_device_event(struct notifier_block *unused, unsigned long event,
 
 		while (mux_dev != NULL) {
 			pdev_priv = MV_MUX_PRIV(mux_dev);
-			if (mux_dev->mtu <= dev->mtu) {
-				mux_dev = pdev_priv->next;
-				continue;
-			}
 			dev_set_mtu(mux_dev, dev->mtu);
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+			if (mux_switch_shadow.attach)
+				mv_mux_switch_mtu_update(mux_dev->mtu);
+#endif
 			mux_dev = pdev_priv->next;
 		}
+
 		break;
 
 	case NETDEV_FEAT_CHANGE:
-- 
1.7.5.4

